[REEF-1004] Define Avro schema for Vortex messages

yunseong opened the pull request:
This addressed the issue by
  * Defining Avro schema for Vortex messages.
  * Adding VortexAvroUtils to serialize/deserialize messages.
  * Replacing SerializationUtils with Avro.
  * Adding TODO comments for user function, input, and output.

JIRA:
  [REEF-1004](https://issues.apache.org/jira/browse/REEF-1004)

Pull Request:
  This closes

yunseong commented:
Oops, I forgot to ask review for this. @bgchun @dafrista Could you take a look? This is similar to how we've done in SNU projects.

yunseong commented:
@bgchun Could you kindly take a look at this, please? Thanks.

bgchun commented on lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/VortexAvroUtils.java:
```diff
@@ -0,0 +1,212 @@
+      final TaskletExecutionRequest taskletExecutionRequest = (TaskletExecutionRequest) vortexRequest;
+      // TODO[REEF-1005]: Allow custom codecs for input/output data in Vortex.
+      final byte[] serializedInput = SerializationUtils.serialize(taskletExecutionRequest.getInput());
+      // TODO[REEF-1003]: Use reflection instead of serialization when launching VortexFunction
```
Can you elaborate?

yunseong replied:
Do you want me to elaborate the TODO comment instead of having just JIRA title? Or do you want me to give some comments of why do we need those TODOs?

bgchun replied:
The latter.

On Thu, Dec 3, 2015 at 4:24 PM, Yunseong Lee <notifications@github.com>
wrote:

> In
> lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/common/VortexAvroUtils.java
> <https://github.com/apache/reef/pull/677#discussion_r46519855>:
>
> > + */
> > +public final class VortexAvroUtils {
> > +  /**
> > +   * Serialize VortexRequest to byte array.
> > +   * @param vortexRequest Vortex request message to serialize.
> > +   * @return Serialized byte array.
> > +   */
> > +  public static byte[] toBytes(final VortexRequest vortexRequest) {
> > +    // Convert VortexRequest message to Avro message.
> > +    final AvroVortexRequest avroVortexRequest;
> > +    switch (vortexRequest.getType()) {
> > +    case ExecuteTasklet:
> > +      final TaskletExecutionRequest taskletExecutionRequest = (TaskletExecutionRequest) vortexRequest;
> > +      // TODO[REEF-1005]: Allow custom codecs for input/output data in Vortex.
> > +      final byte[] serializedInput = SerializationUtils.serialize(taskletExecutionRequest.getInput());
> > +      // TODO[REEF-1003]: Use reflection instead of serialization when launching VortexFunction
>
> Do you want me to elaborate the TODO comment instead of having just JIRA
> title? Or do you want me to give some comments of why do we need those
> TODOs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/reef/pull/677/files#r46519855>.
>



-- 
Byung-Gon Chun

bgchun commented:
@yunseong Overall, this looks good. I do have a question, though. 
Right now, we have only `taskletId`, but a tasklet can be executed multiple times. In other frameworks like Hadoop MR, multiple execution instances are differentiated by an attempt id. For us, this would be `taskletAttemptId`. Don't you think we need attempt ids?

yunseong commented:
@bgchun Thanks for the review! I left a question about your comment.

For `taskletAttemptId`, I think it will be very useful for retrying failed Tasklets or duplicate execution. I'll create a JIRA for that.

yunseong commented:
@bgchun 
* I've added REEF-1038(https://issues.apache.org/jira/browse/REEF-1038) for TaskletAttemptId. 
* I've added comments where you mentioned; there are 3 more places which have `TODO`s, but I did not add comments because the TODO items are same. Please let me know if you think I need elaborate more, or you want me to add comments in other TODOs as well.

bgchun commented:
@yunseong This looks good. I will test and merge.
